tests/filter-video-average.cpp

```diff
@@ -0,0 +1,156 @@
+
+    OK(acquire_configure(runtime, &props));
+
+    AcquirePropertyMetadata metadata = { 0 };
+    OK(acquire_get_configuration_metadata(runtime, &metadata));
+
+    props.video[0].camera.settings.binning = 1;
+    props.video[0].camera.settings.pixel_type = SampleType_u12;
+    props.video[0].camera.settings.shape = {
+        .x = 1920,
+        .y = 1080,
+    };
+    props.video[0].camera.settings.exposure_time_us = 1e5;
+    props.video[0].max_frame_count = 10;
+
+    OK(acquire_configure(runtime, &props));
+
+    const auto next = [](VideoFrame* cur) -> VideoFrame* {
+        return (VideoFrame*)(((uint8_t*)cur) + cur->bytes_of_frame);
+    };
+
+    const auto consumed_bytes = [](const VideoFrame* const cur,
+                                   const VideoFrame* const end) -> size_t {
+        return (uint8_t*)end - (uint8_t*)cur;
+    };
+
+    struct clock clock
+    {};
+    // expected time to acquire frames + 100%
+    static double time_limit_ms = 
+        props.video[0].max_frame_count
+        * (props.video[0].camera.settings.exposure_time_us / 1000.0)
+        * 2.0;
+    clock_init(&clock);
+    clock_shift_ms(&clock, time_limit_ms);
+    OK(acquire_start(runtime));
+    {
+        uint64_t nframes = 0;
+        const uint64_t expected_nframes = props.video[0].max_frame_count / props.video[0].frame_average_count;
+        LOG("Expecting %d frames", expected_nframes);
+        while (nframes < expected_nframes) {
+            struct clock throttle
+            {};
+            clock_init(&throttle);
+            EXPECT(clock_cmp_now(&clock) < 0,
+                   "Timeout at %f ms",
+                   clock_toc_ms(&clock) + time_limit_ms);
+            VideoFrame *beg, *end, *cur;
+            OK(acquire_map_read(runtime, 0, &beg, &end));
+            for (cur = beg; cur < end; cur = next(cur)) {
+                LOG("stream %d counting frame w id %d", 0, cur->frame_id);
+                CHECK(cur->shape.dims.width ==
+                      props.video[0].camera.settings.shape.x);
+                CHECK(cur->shape.dims.height ==
+                      props.video[0].camera.settings.shape.y);
+                ++nframes;
+            }
+            {
+                uint32_t n = (uint32_t)consumed_bytes(beg, end);
+                OK(acquire_unmap_read(runtime, 0, n));
+                if (n)
+                    LOG("stream %d consumed bytes %d", 0, n);
+            }
+            clock_sleep_ms(&throttle, 100.0f);
+
+            LOG("stream %d nframes %d. remaining time %f s",
+                0,
+                nframes,
+                -1e-3 * clock_toc_ms(&clock));
+        }
+
+        // TODO: we want a stronger assertion than this since we want to ensure
+        // the values were averaged correctly.
+        // A simulated camera that outputs constant value frames, where that
+        // constant value increases would be suitable.
+        CHECK(nframes == expected_nframes);
```

Comment:
One idea would be to check that the variance of the uniform random camera has the expected 1/sqrt(n) decay.

Reply:
Ah, nice idea. I think the variance will simply shrink by n where n is the number of frames averaged.